### PR TITLE
enh(chunk): Add docs for chunked upload v2

### PR DESCRIPTION
**Pull request by juliushaertl**
Add documentation for https://github.com/nextcloud/server/pull/27034

### 🖼️ Screenshots

image

### CODE REVIEW

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 1]
```diff
@@ -10,13 +10,113 @@ Uploading large files is always a bit problematic as your connection can be inte
 which will fail your entire upload. Nextcloud has a chunking API where you can
 upload smaller chunks which will be assembled on the server once they are all uploaded.
 
-Usage
------
+There are two versions of the chunking API. Version 1 is the original version and version 2 was built as a backward compatible extension to support uploads directly to supporting target storages like S3. Version 2 is the recommended version to use.
+
+Version 2 comes with a few additional requirements and limitations to consider (compared to version 1):
+- Every request needs to have a `Destination`` header present which specifies the target path of the file
```
Comment (blizzz) [inline developer_manual/client_apis/WebDAV/chunking.rst:16]:
```suggestion
- Every request needs to have a `Destination` header present which specifies the target path of the file
```

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 2]
```diff
@@ -10,13 +10,113 @@ Uploading large files is always a bit problematic as your connection can be inte
...
+
+Start by creating a folder with a unique name. You can list the current available
+folder but if you take a random UUID chances of collision are tiny.
+
+.. code-block:: console
+
+    curl -X MKCOL -u roeland:pass \
+        https://server/remote.php/dav/uploads/roeland/myapp-e1663913-4423-4efe-a9cd-26e7beeca3c0 \
+        --header 'Destination: Destination
+	https://server/remote.php/dav/files/roeland/dest/file.zip'
```
Comment (blizzz) [inline developer_manual/client_apis/WebDAV/chunking.rst:45]: not an error, but the line break is confusing. would expect it ending at the same line, or having further indentation.

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 3]
```diff
@@ -10,13 +10,113 @@ Uploading large files is always a bit problematic as your connection can be inte
...
+- The size of chunks must be between 5MB and 5GB (except for the last chunk, which can be smaller)
+
+
+.. code-block:: console
+
+    curl -X PUT -u roeland:pass \
+        https://server/remote.php/dav/uploads/roeland/myapp-e1663913-4423-4efe-a9cd-26e7beeca3c0/00001 \
+        --data-binary @chunk1 \
+        --header 'Destination: Destination
+	https://server/remote.php/dav/files/roeland/dest/file.zip'
```
Comment (blizzz) [inline developer_manual/client_apis/WebDAV/chunking.rst:63]: see above

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 4]
```diff
@@ -10,13 +10,113 @@ Uploading large files is always a bit problematic as your connection can be inte
...
+
+    curl -X PUT -u roeland:pass \
+        https://server/remote.php/dav/uploads/roeland/myapp-e1663913-4423-4efe-a9cd-26e7beeca3c0/00002 \
+        --data-binary @chunk2 \
+        --header 'Destination: Destination
+	https://server/remote.php/dav/files/roeland/dest/file.zip'
```
Comment (blizzz) [inline developer_manual/client_apis/WebDAV/chunking.rst:69]: see above

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 5]
```diff
@@ -10,13 +10,113 @@ Uploading large files is always a bit problematic as your connection can be inte
...
+
+Assembling the chunks
+^^^^^^^^^^^^^^^^^^^^^
+
+Assembling the chunk on the server is a matter of initiating a move from the client.
+
+.. code-block:: console
+
+    curl -X MOVE -u roeland:pass
+        --header 'Destination:https://server/remote.php/dav/files/roeland/dest/file.zip' \
```
Comment (blizzz) [inline developer_manual/client_apis/WebDAV/chunking.rst:82]:
```suggestion
        --header 'Destination: https://server/remote.php/dav/files/roeland/dest/file.zip' \
```

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 6]
```diff
@@ -10,13 +10,113 @@ Uploading large files is always a bit problematic as your connection can be inte
...
+
+The server will now assemble the chunks and move the final file to the folder ``dest/file.zip``.
+
+If a modification time should be set, you can by adding it as header with date as unixtime:
+
+.. code-block:: console
+
+    curl -X MOVE -u roeland:pass
+        --header 'Destination:https://server/remote.php/dav/files/roeland/dest/file.zip' \
+        --header 'X-OC-Mtime:1547545326' \
```
Comment (blizzz) [inline developer_manual/client_apis/WebDAV/chunking.rst:93]:
```suggestion
        --header 'Destination: https://server/remote.php/dav/files/roeland/dest/file.zip' \
        --header 'X-OC-Mtime: 1547545326' \
```

**Review by blizzz — COMMENTED** [review]
formatting

File: developer_manual/client_apis/WebDAV/chunking.rst [hunk 7]
```diff
@@ -60,7 +158,7 @@ Assembling the chunk on the server is a matter of initiating a move from the cli
 The server will now assemble the chunks and move the final file to the folder ``dest/file.zip``.
 
 If a modification time should be set, you can by adding it as header with date as unixtime:
-``curl -X MOVE -u roeland:pass --header 'X-OC-Mtime:1547545326' --header 'Destination:https://server/remote.php/dav/files/roeland/dest/file.zip' https://server/remote.php/dav/uploads/roeland/myapp-e1663913-4423-4efe-a9cd-26e7beeca3c0/.file``" 
+``curl -X MOVE -u roeland:pass --header 'X-OC-Mtime:1547545326' --header 'Destination:https://server/remote.php/dav/files/roeland/dest/file.zip' https://server/remote.php/dav/uploads/roeland/myapp-e1663913-4423-4efe-a9cd-26e7beeca3c0/.file``"
 Otherwise the current upload date will be used as modification date.
```
Comment (tflidd) [inline developer_manual/client_apis/WebDAV/chunking.rst:162]: The upload date is the date of the first/last file uploaded or when the `MOVE` command is passed?

**Comment by tflidd** [comment]
Nice comprehensible write-up with example code for this feature :+1:

What is not clear, if your connection is interrupted and you want to continue to upload the data, how long will Nextcloud keep the chunks? There probably is some clean up routine. I imagine you can just check if the upload folder is still present (what was mentioned in the beginning). Perhaps too obvious?

**Comment by juliushaertl** [comment]
> What is not clear, if your connection is interrupted and you want to continue to upload the data, how long will Nextcloud keep the chunks? There probably is some clean up routine. I imagine you can just check if the upload folder is still present (what was mentioned in the beginning). Perhaps too obvious?

I'll clarify that, there is actually also depend on the storage (e.g. s3) policies how long the multipart uploads are kept. On the Nextcloud side we expire them [after 24 hours](https://github.com/nextcloud/server/blob/944675a4a2cf95d9b2c71f448cffd75c75829e8c/apps/dav/lib/BackgroundJob/UploadCleanup.php#L74)

**Review by mpivchev — APPROVED** [review]

**Comment by juliushaertl** [comment]
Added a note about expiring upload folders

**Pull request merged**